### PR TITLE
KBV-179 - Restructured the project based on the address CRI structure

### PR DESCRIPTION
**Pull request by amanm-alphagov**
Defined KBV API by using an OpenAPI 3.0 definition file in the `api.yaml` file. This is now imported in`template.yaml` and forms the API Gateway.

Project restructuring changes have already been completed earlier by Ola.

### What changed

The `question` endpoint although is a GET method but needs to have it set as POST in the `x-amazon-apigateway-integration` object. More details can be found [here](https://aws.amazon.com/premiumsupport/knowledge-center/api-gateway-lambda-template-invoke-error/).

### CODE REVIEW

File: deploy/template.yaml [hunk 1]
```diff
@@ -126,7 +110,7 @@ Resources:
           EXPERIAN_API_WRAPPER_SAA_RESOURCE: /question-request
           EXPERIAN_API_WRAPPER_URL: https://zn3dus332j.execute-api.eu-west-2.amazonaws.com/dev
```
Comment (KaraokeStu) [inline deploy/template.yaml:111]: This should ideally be passed as a parameter, either from the parameter store, or via the samconfig.toml file

**Review by KaraokeStu — CHANGES_REQUESTED** [review]
We should also use AWS Powertools for logging (see the Address CRI)

File: deploy/template.yaml (hunk 1 above)
Comment (amanm-alphagov) [reply deploy/template.yaml:111]: Agreed. It is part of another story - https://govukverify.atlassian.net/browse/KBV-245

File: deploy/template.yaml [hunk 2]
```diff
@@ -31,35 +23,33 @@ Parameters:
...
     ConstraintDescription: specify dev, staging, integration or prod for environment
 
-
 Resources:
 
   KBVSessionApi:
     Type: AWS::Serverless::Api
     Properties:
+      MethodSettings:
+        - LoggingLevel: INFO
```
[the remark this answers is not in the document]
Comment (amanm-alphagov) [reply deploy/template.yaml:32]: @KaraokeStu if I understand this correctly, the logging should be at global level in the template file? Currently it is at the REST API level and not at the function level.

File: deploy/api.yaml [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+      type: "object"
+      properties:
+        questionId:
+          description: The unique identifier for the question
+          maxLength: 6
+          type: "string"
+        answer:
+          description: Answer provided by the user
+          maxLength: 50
+          type: "string"
```
Comment (jkunle) [inline deploy/api.yaml:157]: Please add a newline here

**Comment by amanm-alphagov** [comment]
> We should also use AWS Powertools for logging (see the Address CRI)

Implemented as part of PR - https://github.com/alphagov/di-ipv-cri-kbv-api/pull/18

**Review by jkunle — DISMISSED** [review]
This can be merged and issues addressed in the next PR

File: deploy/template.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (amanm-alphagov) [reply deploy/template.yaml:36]: added powertools at global level

File: deploy/api.yaml [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+      parameters:
+        - $ref: "#/components/parameters/SessionHeader"
+      responses:
+        "200":
+          description: "200 response"
+          content:
+            application/json:
+              schema:
+                $ref: "#/components/schemas/Question"
+        "201":
```
Comment (merlinc) [inline deploy/api.yaml:46]: The original diagram was incorrect - this status code should be `204 - no content` not `201 - created`

File: deploy/api.yaml [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+  schemas:
+    Session:
+      type: "object"
+      properties:
+        session-id:
+          type: "string"
+    Error:
+      title: "Error Schema"
+      type: "object"
+      properties:
```
Comment (merlinc) [inline deploy/api.yaml:119]: We should return to this when we look at the error flows, as we the frontend will almost always be redirecting back to core as part of the oauth flow, and the core system rendering the error pages.

As this redirect will have an error code in the param, as always we'll want to balance usability with security.

File: deploy/api.yaml [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+    Question:
+      type: "object"
+      properties:
+        questionID:
+          type: "string"
+        text:
+          type: "string"
+        tooltip:
+          type: "string"
+        answerHeldFlag:
```
Comment (merlinc) [inline deploy/api.yaml:131]: What needs this property?

File: deploy/api.yaml [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+          items:
+            type: "string"
+    Answer:
+      type: "object"
+      properties:
+        questionId:
+          description: The unique identifier for the question
+          maxLength: 6
+          type: "string"
+        answer:
```
Comment (merlinc) [inline deploy/api.yaml:154]: For the benefit of people not familiar with the Experian API, answers are returned as text, without an ID, which is why the `Answer` has a `questionId` but a `answer` string.

File: deploy/api.yaml [hunk 8]
```diff
@@ -0,0 +1,157 @@
...
+      in: header
+      description: A UUID generated by the KBV API to act as a primary key for the KBVSessionTable in DynamoDB
+      required: true
+      schema:
+        type: string
+  schemas:
+    Session:
+      type: "object"
+      properties:
+        session-id:
```
Comment (merlinc) [inline deploy/api.yaml:114]: We use `kebab-case` for `session-id` and `lowerCamelCase` for `questionId`.

Which one should we use for consistency?

**Review by merlinc — DISMISSED** [review]
I've reviewed the API gateway as that's what I'll be consuming.

All the comments on errors are to keep them in our mind for when we refactor this area.

File: deploy/api.yaml (hunk 6 above)
Comment (amanm-alphagov) [reply deploy/api.yaml:131]: The property is sent back in the `Question` payload we receive from the Experian API. It is currently part of the payload and I have not really dug deep into what is its exact need. We can ask the API to ignore it but we may need it in perhaps in the future flows.

File: deploy/api.yaml (hunk 7 above)
Comment (amanm-alphagov) [reply deploy/api.yaml:154]: I can confirm the above statement is correct.

File: deploy/api.yaml (hunk 8 above)
Comment (amanm-alphagov) [reply deploy/api.yaml:114]: I think they are both different objects of Http request. `session-id` is part of the `request header` and `questionId` is part of the POST `request body` payload. 
When we are using the request body object (questionId) and it is a Java object which will be marshalled during conversion to a JSON object. Java variables cannot be declared with a hyphen (-). We use lowerCamelCase for variable declaration in many places.

File: deploy/api.yaml (hunk 5 above)
Comment (amanm-alphagov) [reply deploy/api.yaml:119]: Agreed

**Review by jkunle — APPROVED** [review]

**Pull request merged**